crates/router/src/core/payments/payment_methods.rs

```diff
@@ -0,0 +1,270 @@
+
+use common_utils::{ext_traits::OptionExt, id_type};
+use error_stack::ResultExt;
+
+use super::errors;
+use crate::{db::errors::StorageErrorExt, routes, types::domain};
+
+#[cfg(all(
+    feature = "v2",
+    feature = "customer_v2",
+    feature = "payment_methods_v2"
+))]
+pub async fn list_payment_methods(
+    state: routes::SessionState,
+    merchant_account: domain::MerchantAccount,
+    profile: domain::Profile,
+    key_store: domain::MerchantKeyStore,
+    payment_id: id_type::GlobalPaymentId,
+    _req: api_models::payments::PaymentMethodsListRequest,
+    header_payload: &hyperswitch_domain_models::payments::HeaderPayload,
+) -> errors::RouterResponse<api_models::payments::PaymentMethodListResponseForPayments> {
+    let db = &*state.store;
+    let key_manager_state = &(&state).into();
+
+    let payment_intent = db
+        .find_payment_intent_by_id(
+            key_manager_state,
+            &payment_id,
+            &key_store,
+            merchant_account.storage_scheme,
+        )
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+
+    validate_payment_status(payment_intent.status)?;
+
+    let client_secret = header_payload
+        .client_secret
+        .as_ref()
+        .get_required_value("client_secret header")
+        .change_context(errors::ApiErrorResponse::MissingRequiredField {
+            field_name: "client_secret header",
+        })?;
+
+    payment_intent.validate_client_secret(client_secret)?;
+
+    let payment_connector_accounts = db
+        .list_enabled_connector_accounts_by_profile_id(
+            key_manager_state,
+            profile.get_id(),
+            &key_store,
+            common_enums::ConnectorType::PaymentProcessor,
+        )
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("error when fetching merchant connector accounts")?;
+
+    let response = PaymentMethodsEnabled::from_payment_connectors_list(payment_connector_accounts)
+        .perform_filtering()
+        .get_required_fields(RequiredFieldsInput::new())
+        .perform_surcharge_calculation()
+        .generate_response();
+
+    Ok(hyperswitch_domain_models::api::ApplicationResponse::Json(
+        response,
+    ))
+}
+
+/// Holds the payment methods enabled for a connector along with the connector name
+/// This struct is a flattened representation of the payment methods enabled for a connector
+#[derive(Debug)]
+struct PaymentMethodsEnabledForConnector {
+    payment_methods_enabled: common_utils::types::RequestPaymentMethodTypes,
+    payment_method: common_enums::PaymentMethod,
+    connector: String,
+}
+
+/// Container for the inputs required for the required fields
+struct RequiredFieldsInput {}
+
+impl RequiredFieldsInput {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+/// Container for the filtered payment methods
+struct FilteredPaymentMethodsEnabled {
```

Comment:
The underlying type is Vec<PaymentMethodsEnabledForConnector> so this struct we can name as PaymentMethodsEnabledForConnectors